Maven 3 task: JAVA_HOME is not correctly set.

We have a problem with Artifactory Maven 3 task and maven-javadoc-plugin:
> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.1.1:jar (attach-javadocs) on project: MavenReportException: Error while creating archive: Unable to find javadoc command: The environment variable JAVA_HOME is not correctly set.

Looks like Bamboo Artifactory Plugin doesn't set JAVA_HOME 

There is a point about it on http://maven.apache.org/install.html
> Ensure JAVA_HOME environment variable is set and points to your JDK installation

`artifactory.task.override.jdk` isn't working too (https://www.jfrog.com/confluence/pages/viewpage.action?pageId=46108401#BambooArtifactoryPlug-in-ControllingtheBuildJDKusingBambooVariables)
I tried `artifactory.task.override.jdk.env.var=bamboo_capability_system_jdk_JDK_11` and `artifactory.task.override.jdk.env.var=bamboo.capability.system.jdk.JDK-11`
> The task is configured to use the 'bamboo_capability_system_jdk_JDK_11' environment variable for the build JDK, but this environment variable is not defined.`

I also tried `builder.artifactoryMaven3Builder.environmentVariables="JAVA_HOME=${bamboo_capability_system_jdk_JDK_11}"`
Doesn't work. `bamboo_capability_system_jdk_JDK_11` and `bamboo.capability.system.jdk.JDK-11`
> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.1.1:jar (attach-javadocs) on project: MavenReportException: Error while generating Javadoc: Unable to find javadoc command: The environment variable JAVA_HOME=${bamboo_capability_system_jdk_JDK_11} doesn't exist or is not a valid directory.`

So we only can hardcode JDK-path in `builder.artifactoryMaven3Builder.environmentVariables="JAVA_HOME=/path/to/jdk"` or make changes in pom.xml (https://stackoverflow.com/a/54323506) for all projects

Meanwhile, the default Maven 3 plugin (https://docs.atlassian.com/bamboo/docs-0606/Maven) works fine.

--- a/src/main/java/org/jfrog/bamboo/task/ArtifactoryMaven3Task.java
+++ b/src/main/java/org/jfrog/bamboo/task/ArtifactoryMaven3Task.java
@@ -115,6 +115,10 @@ public class ArtifactoryMaven3Task extends BaseJavaBuildTask {
         }
         List<String> command = buildCommand(mavenHome, rootDirectory, systemProps);
 
+        // Override the JAVA_HOME according to the build configuration.
+        String jdkPath = getConfiguredJdkPath(buildParamsOverrideManager, mavenBuildContext, capabilityContext);
+        environmentVariables.put("JAVA_HOME", jdkPath);
+
         ExternalProcess process = getExternalProcess(taskContext, rootDirectory, command, environmentVariables);
         try {
             executeExternalProcess(logger, process, log);

--- a/src/main/java/org/jfrog/bamboo/task/ArtifactoryMaven3Task.java
+++ b/src/main/java/org/jfrog/bamboo/task/ArtifactoryMaven3Task.java
@@ -19,7 +19,6 @@ import org.jfrog.bamboo.admin.ServerConfig;
 import org.jfrog.bamboo.builder.BuilderDependencyHelper;
 import org.jfrog.bamboo.builder.MavenDataHelper;
 import org.jfrog.bamboo.context.Maven3BuildContext;
-import org.jfrog.bamboo.context.PackageManagersContext;
 import org.jfrog.bamboo.util.PluginProperties;
 import org.jfrog.bamboo.util.TaskUtils;
 import org.jfrog.bamboo.util.Utils;
@@ -113,10 +112,10 @@ public class ArtifactoryMaven3Task extends BaseJavaBuildTask {
             log.error(logger.addErrorLogEntry("Maven home is not defined!"));
             return TaskResultBuilder.newBuilder(taskContext).failed().build();
         }
-        List<String> command = buildCommand(mavenHome, rootDirectory, systemProps);
+        String jdkPath = getConfiguredJdkPath(buildParamsOverrideManager, mavenBuildContext, capabilityContext);
+        List<String> command = buildCommand(mavenHome, jdkPath, rootDirectory, systemProps);
 
         // Override the JAVA_HOME according to the build configuration.
-        String jdkPath = getConfiguredJdkPath(buildParamsOverrideManager, mavenBuildContext, capabilityContext);
         environmentVariables.put("JAVA_HOME", jdkPath);
 
         ExternalProcess process = getExternalProcess(taskContext, rootDirectory, command, environmentVariables);
@@ -150,8 +149,7 @@ public class ArtifactoryMaven3Task extends BaseJavaBuildTask {
      *
      * @return Java bin path
      */
-    public String getJavaExecutable(PackageManagersContext context) throws TaskException {
-        String jdkPath = getConfiguredJdkPath(buildParamsOverrideManager, context, capabilityContext);
+    public String getJavaExecutable(String jdkPath) {
         StringBuilder binPathBuilder = new StringBuilder(jdkPath);
         if (SystemUtils.IS_OS_WINDOWS) {
             binPathBuilder.append("bin").append(File.separator).append("java.exe");
@@ -170,8 +168,8 @@ public class ArtifactoryMaven3Task extends BaseJavaBuildTask {
         return binPath;
     }
 
-    private List<String> buildCommand(String mavenHome, File rootDirectory, List<String> systemProps) throws TaskException {
-        List<String> command = getCommand(mavenBuildContext);
+    private List<String> buildCommand(String mavenHome, String jdkPath, File rootDirectory, List<String> systemProps) throws TaskException {
+        List<String> command = getCommand(jdkPath);
         appendClassPathArguments(command, mavenHome);
         appendClassWorldsConfArgument(command, mavenHome);
         appendBuildInfoPropertiesArgument(command);
@@ -200,9 +198,9 @@ public class ArtifactoryMaven3Task extends BaseJavaBuildTask {
         command.add(Commandline.quoteArgument("-Dmaven.multiModuleProjectDirectory" + "=" + rootDirectory.getPath()));
     }
 
-    private List<String> getCommand(Maven3BuildContext mavenBuildContext) throws TaskException {
+    private List<String> getCommand(String jdkPath) throws TaskException {
         List<String> command = Lists.newArrayList();
-        String executable = getJavaExecutable(mavenBuildContext);
+        String executable = getJavaExecutable(jdkPath);
         if (StringUtils.isBlank(executable)) {
             log.error("No Maven executable found");
             return command;

